File: leetcode-scala/src/dp/easy/RangeSumQueryImmutable.scala
package dp.easy

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/8/28 11:25
 * @Description RangeSumQueryImmutable
 */
object RangeSumQueryImmutable {

    class NumArray(_nums: Array[Int]) {

        val nums: Array[Int] = _nums

        def sumRange(i: Int, j: Int): Int = {
            val sums = nums.scanLeft(0)(_ + _)
            sums(j + 1) - sums(i)
        }
    }

    def main(args: Array[String]): Unit = {
        println(new NumArray(Array.range(1, 10)).sumRange(0, 3))
    }
}

File: leetcode-java/src/classify/dp/RangeSumQueryImmutable.java
package classify.dp;

/**
 * @author yutiantang
 * @create 2021/5/12 21:16
 */
public class RangeSumQueryImmutable {

    /**
     * 13ms     37.14%
     * 43.2MB   5.14%
     */
    static class NumArray {

        private final int[] sums;

        public NumArray(int[] nums) {
            sums = new int[nums.length];
            sums[0] = nums[0];
            for (int i = 1; i < nums.length; i++) {
                sums[i] = sums[i - 1] + nums[i];
            }
        }

        public int sumRange(int left, int right) {
            return left == 0 ? sums[right] : sums[right] - sums[left - 1];
        }
    }
}
